xdsl/transforms/printf_to_llvm.py

```diff
@@ -145,9 +146,21 @@ class PrintfToLLVM(ModulePass):
     name = "printf-to-llvm"
 
     def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        has_done_action = False
+
+        def handler(op: Operation, new_results: Sequence[SSAValue | None]):
+            nonlocal has_done_action
+            has_done_action = True
```

Comment:
This technically works, but I would instead override the `operation_insertion_handler`, and check if one with the `CallOp("printf")` is being inserted.
Just to make this future-proof and avoiding bug.

I would also rename `has_done_action` with something else, just to make it clear why we have this variable. Something like "has_inserted_printf_call"?

Reply:
The rewrite patterns are ideally stateless. Having state in them makes it much more difficult to reason about. I know that we currently don't have this as a strict restriction, but it would be good not to add extra state to existing patterns if we can avoid it. To me it makes intuitive sense that the Rewriter should track whether it's done an action, as opposed to a pattern.